pallets/relay-storage-roots/src/lib.rs

```diff
@@ -0,0 +1,117 @@
+// Copyright Moonsong Labs
+// This file is part of Moonkit.
+
+// Moonkit is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Moonkit is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Moonkit.  If not, see <http://www.gnu.org/licenses/>.
+
+//! # Relay Storage Roots Pallet
+//!
+//! This pallet stores the latest `MaxStorageRoots` relay storage roots, which can be used to
+//! verify state proofs against an old state of the relay chain.
+//! 
+//! This is useful when the proof needs to be generated by an end user, because
+//! by the time their transaction is included in a block, the latest relay
+//! block will probably have changed and therefore the proof will be invalid.
+//! To avoid that, we expect the user to generate a proof against the latest relay block stored
+//! in this pallet. This proof will then be valid as long as that relay block is not removed from
+//! here.
+//! 
+//! This pallet SHOULD NOT be used for data that can change quickly, because we allow the user to
+//! submit a proof of an old state. Therefore a valid proof does not imply that the current relay
+//! state is the expected one.
+
+#![cfg_attr(not(feature = "std"), no_std)]
+
+pub use crate::weights::WeightInfo;
+use cumulus_pallet_parachain_system::RelaychainStateProvider;
+use cumulus_primitives_core::relay_chain::BlockNumber as RelayBlockNumber;
+use frame_support::pallet;
+use frame_support::pallet_prelude::*;
+use frame_system::pallet_prelude::*;
+pub use pallet::*;
+use sp_core::Get;
+use sp_core::H256;
+use sp_std::collections::vec_deque::VecDeque;
+
+#[cfg(any(test, feature = "runtime-benchmarks"))]
+mod benchmarks;
+pub mod weights;
+
+#[cfg(test)]
+mod mock;
+#[cfg(test)]
+mod tests;
+
+#[pallet]
+pub mod pallet {
+	use super::*;
+
+	#[pallet::pallet]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(PhantomData<T>);
+
+	/// Configuration trait of this pallet.
+	#[pallet::config]
+	pub trait Config: frame_system::Config {
+		type RelaychainStateProvider: RelaychainStateProvider;
+		#[pallet::constant]
+		type MaxStorageRoots: Get<u32>;
+		/// Weight info
+		type WeightInfo: WeightInfo;
+	}
+
+	/// Map of relay block number to relay storage root
+	#[pallet::storage]
+	pub type RelayStorageRoot<T: Config> =
+		StorageMap<_, Twox64Concat, RelayBlockNumber, H256, OptionQuery>;
+
+	/// List of all the keys in `RelayStorageRoot`.
+	/// Used to remove the oldest key without having to iterate over all of them.
+	#[pallet::storage]
+	pub type RelayStorageRootKeys<T: Config> =
```

Comment:
I also attempted this, the problem is that the drain iterator is not sorted. So if I want to remove the oldest key, I need to iterate over all of them. Removing the first element of the iterator will remove the element whose hash is the lowest. Actually now I see the docs mention this:

> Remove all elements from the map and iterate through them in lexicographical order of the encoded key.

Reply:
Maybe it will work if we set the hash to identity. Do you think that's a good idea?

Reply:
@tmpolaczyk as long as MaxStorageRoots is not too high (less than 1000) I think that the Identity hash is safe.

Reply:
Ok we expect this value to be around 10 so I implemented this change. I am not sure if I like it, so I did it in a separate commit that is easy to revert in case we change our minds.

Reply:
Maybe something we can do is make sure we dont go over with integrity-test hook?

Reply:
We can also define an hardcoded constant in the pallet and use `max(MaxStorageRoots, 1_000)`, this way the pallet itsefl ensure that the size of the map is reasonable

Reply:
yeah that's also true. I like this better

Reply:
@librelois I just found out that the keys are stored in little endian, so the identity hash doesn't work with a key of more than 1 byte... We could solve that using a "reverse identity" hash, or by using a custom type as the key, but after discussing with @girazoki we believe it would be better to use the old approach with `RelayStorageRootKeys`.

Reply:
I very don't like the old approach because you need to decode and proccess the whole VecDeQueue. In that case it's better to store the whole map in one storage value directly, but it doesn't scale.

A better approach can be to add a storage value `OldestRelayNumber` to keep track what is the next key to remove without iterating

Reply:
In that case we would need a linked list to update `OldestRelayNumber`, so the map value would need to be:

```
struct MapValue {
    relay_storage_root: H256,
    next_key: u32,
}
```

And the insert operation would also be more complex, because we would need to keep track of the newest key and update its `next_key`. So I prefer the `VecDeque`.

Reply:
I don't understand why you see the need for a linked list.

We just want to keep the last K relay hash, don't we?
So you just need to increment `OldestRelayNumber` to `CurrentRelayNumber-K` and delete the corrresponding key at each increment.

Reply:
In any case, `VecDeque` is a bad approach becasue it add more complex storage for nothing. At worst, store a `BTreeMap<u32, Hash>` directly in one StorageValue and for K ~ 10 it will work.

Reply:
> And in the case of parathreads there may be huge gaps.

Sorry, I didn't think about parathreads, it's true that for parathreads the gap can be huge and we can't just increment.

Thanks for the explanation, I understand better the relevance of your initial design now, but at first glance it's not obvious, can you please add a comment in the code that explains the  rationale for this design choice?

Reply:
Sure, I added some docs. Also changed the storage item to use a `BoundedVec`, which is converted into a `VecDeque` after reading it.